More networking tests

danwinship opened the pull request:
Makes the isolation test get run under both plugins (testing that things *aren't* isolated when using the flat plugin), and adds a basic service test and a service isolation test.

@marun PTAL

marun commented on test/extended/networking/isolation.go:
```diff
@@ -7,13 +7,17 @@ import (
 
-	It("should prevent communication between pods in different namespaces", func() {
+	var expectation string
+	if isMultitenant() {
```
I think it would be preferable to have the test methods defined statically and skipped as appropriate.  It might be confusing to those analyzing test results if the list of tests varied according to the setup being tested. Consider instead creating a parametrized helper method that can be called from 2 tests, e.g.

    multitenant := isMultitenant()
    It("should prevent...") {
        if multitenant {
            checkPodIsolation(multitenant)
        } else {
            e2e.Skipf("Tenant isolation is not supported")
        }
    }
    It("should allow...") {
        ...
    }

danwinship replied:
Trying to do this results in:

    /data/src/github.com/openshift/origin/test/extended/networking/isolation.go:58
    
    You may only call BeforeEach from within a Describe or Context

(I'm not calling BeforeEach at all.) It doesn't seem to like moving By() calls into a separate function from the It()/Describe().

marun replied:
Fair enough.  I'd suggest ditching the 'By' comments then and relying instead on informational logging, which is all 'By' really is anyway.

danwinship replied:
my bad, it was the `e2e.NewFramework()` calls that were the problem, and I can just leave those outside the helper

marun commented on test/extended/networking/isolation.go:
```diff
@@ -24,12 +28,20 @@ var _ = Describe("[networking] network isolation plugin", func() {
 		ip := launchWebserverPod(f1, podName, node1.Name)
 
-		By("Checking that the webserver is not accessible from a pod in a different namespace on the same node")
+		By("Trying to connect to the webserver from a pod in a different namespace on the same node")
```
Consider separating the single node and multinode checks into separate tests so that isolation can be checked in a single node configuration and multinode isolation can be skipped if only a single node is available.  This is the pattern of the upstream version of the sanity checks that will eventually replace the checks in origin:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/networking.go#L216

marun commented on test/extended/networking/util.go:
```diff
@@ -211,3 +247,7 @@ func checkConnectivityToHost(f *e2e.Framework, nodeName string, podName string,
 }
+
+func isMultitenant() bool {
+	return os.Getenv("OPENSHIFT_NETWORK_PLUGIN") == "redhat/openshift-ovs-multitenant"
```
Consider setting a separate flag to indicate whether namespaces should be isolated from a networking perspective.  There is ongoing work to make the e2e tests consumable downstream and it may make sense to allow other plugins to be able to run the isolation tests.  It also might make more sense if isolation checks are ever submitted upstream.

marun commented on test/extended/networking/util.go:
```diff
@@ -158,7 +159,10 @@ func launchWebserverPod(f *e2e.Framework, podName string, nodeName string) (ip s
 	pod := &api.Pod{
 		ObjectMeta: api.ObjectMeta{
-			Name: podName,
+			Name:   podName,
```
Consider applying this change upstream also so that kube's LaunchWebserverPod method can replace origin's version:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/util.go#L3504

marun commented on test/extended/networking/util.go:
```diff
@@ -184,6 +188,38 @@ func launchWebserverPod(f *e2e.Framework, podName string, nodeName string) (ip s
 	return
 }
 
+func launchWebserverService(f *e2e.Framework, serviceName string, nodeName string) (serviceIP string) {
```
Consider renaming the return value to serviceSocket, since 'socket' is the combination of hostname/ip:port according to rfc793.

danwinship replied:
I left it as "serviceIP" because that matches LaunchWebserverPod()

marun replied:
Arg, I somehow missed that.  -1 on me for making the original mistake, and I would encourage you to avoid repeating my mistakes.

marun commented on test/extended/networking/services.go:
```diff
@@ -0,0 +1,67 @@
+	})
+
+	var expectation string
+	if isMultitenant() {
```
Same comment here as the pod isolation check regarding the desirability of defining tests statically.

marun commented on test/extended/networking/services.go:
```diff
@@ -0,0 +1,67 @@
+	f1 := e2e.NewFramework("net-isolation1")
+	f2 := e2e.NewFramework("net-isolation2")
+
+	It("should allow connections to other pods via service IPs", func() {
```
Consider skipping this test instead of failing (via getMultipleNodes) if only one node is found to be present.  The kube e2e service conformance check will effectively validate single node service connectivity in that case:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/service.go#L82

marun commented on test/extended/networking/services.go:
```diff
@@ -0,0 +1,67 @@
+		defer f1.Client.Services(f1.Namespace.Name).Delete(serviceName)
+		ip := launchWebserverService(f1, serviceName, node1.Name)
+
+		By("Trying to connect to the service from a pod in a different namespace on the same node")
```
Same comment here as the pod isolation checks regarding the desirability of separating single and multinode checks.

marun commented:
Finally, more networking tests!  Comments inline are largely structural, you've done a good job here already.

marun commented:
[testonlyextended][extended:networking]

danwinship commented:
Updated for all comments except as noted above. (Also, some notes in the commit messages.)

I don't really like splitting the isolation tests into separate prevent/allow tests and then always skipping one of them. It feels weird... What do you think about having the list of tests be static but the expectations change? Eg:

    It("handles communication between pods in different namespaces on the same pod in the expected way for this plugin type", func () {
            err := checkPodIsolation(f1, f2, false)
            if pluginIsolatesNamespaces() {
                    Expect(err).To(HaveOccurred())
            } else {
                    Expect(err).To(BeNil())
            }
    })

(Maybe with an extra `e2e.Logf()` in there to clarify "the expected way")

danwinship commented:
[testonlyextended][extended:networking]

danwinship commented:
> I don't really like splitting the isolation tests into separate prevent/allow tests and then always skipping one of them. It feels weird... What do you think about having the list of tests be static but the expectations change?

Or maybe the split+skip approach would seem less silly if the conditions explicitly noted that they were conditional:

    Specify("when using a multi-tenant plugin, it should prevent communication ...
            if !pluginIsolatesNamespaces() {
                    e2e.Skipf("not a multi-tenant plugin")

marun commented:
@danwinship I think tests need to be consistent in their evaluation, and one of the reasons that test skipping is a common pattern.  If more descriptive test names are what it takes to make split+skip palatable to you, though, please feel free.

marun commented on test/extended/networking/isolation.go:
```diff
@@ -1,35 +1,74 @@
-		node1 := nodes.Items[0]
-		node2 := nodes.Items[1]
+	It("should prevent communication between pods in different namespaces on the same node", func() {
+		if pluginIsolatesNamespaces() {
```
Consider assigning the result of this call to a variable and using it place of the 4 places the method is called.

danwinship replied:
I ended up adding helper methods as you suggested later, which means we're still calling it 4 times, but there's no obvious place to cache the return value. But I didn't really care about this anyway: the gratuitous extra calls to pluginIsolatesNamespaces() are orders of magnitude faster than the actual tests anyway

marun replied:
Fair enough.  It wasn't the function call overhead I was worried about, more DRY.

marun commented on test/extended/networking/isolation.go:
```diff
@@ -1,35 +1,74 @@
+		if pluginIsolatesNamespaces() {
+			checkPodIsolation(f1, f2, false)
+		} else {
+			e2e.Skipf("Network plugin does not implement isolation.")
```
Consider defining constants (or a helper method) for implements/does not implement skip.

marun commented on test/extended/networking/isolation.go:
```diff
@@ -1,35 +1,74 @@
+	ip := e2e.LaunchWebserverPod(f1, podName, serverNode.Name)
+
+	err = checkConnectivityToHost(f2, clientNode.Name, "isolation-wget", ip, 10)
+	if pluginIsolatesNamespaces() {
```
Consider passing this value in to the function to avoid unnecessary coupling to the test environment.

marun commented on test/extended/networking/services.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+var _ = Describe("[networking] services", func() {
+	f1 := e2e.NewFramework("net-isolation1")
```
Consider varying the framework names from those used in isolation.go to simplify debugging.

marun commented on test/extended/networking/services.go:
```diff
@@ -0,0 +1,79 @@
+	})
+
+	It("should allow connections to another pod in a different namespace on the same node via a service IP", func() {
+		if pluginIsolatesNamespaces() {
```
Same suggestion regarding using a variable in place of calling this function in each test.

marun commented on test/extended/networking/services.go:
```diff
@@ -0,0 +1,79 @@
+	})
+})
+
+func checkService(serverFramework, clientFramework *e2e.Framework, differentNodes bool) error {
```
Consider using a more descriptive name (e.g. checkServiceIsolation for consistency with checkPodIsolation).

danwinship replied:
The function is used for both isolated and non-isolated (basic sanity-checking) service testing

marun replied:
So maybe 'checkServiceConnectivity'?  I think it helps to have function names be self-documenting, and 'checkService' doesn't really tell me anything.

marun commented on test/extended/networking/util.go:
```diff
@@ -112,75 +97,42 @@ func providerIs(providers ...string) bool {
+	pod, err := podClient.Get(serviceName)
+	expectNoError(err)
+	pod.ObjectMeta.Labels = make(map[string]string)
+	pod.ObjectMeta.Labels["name"] = "web"
```
Consider defining the service first and copying the label to the pod rather than defining independently.

marun commented:
All comments save one are nitty suggestions (varying the framework name in services.go).  Please take the others with a grain of salt, as I'm just being thorough/pedantic.  You've done a good job of cleaning things up.

Any interest in proposing this upstream?  Given a default 'is isolated' of false, the tests that would run would be validating default kube behavior.  Plugin implementers would be able to enable isolation checks if that matched their plugin's capability.

danwinship commented:
> Any interest in proposing this upstream?

Do you have any thoughts on how we ought to be integrating with the existing "provider" stuff? There are a bunch of upstream e2e network tests that are getting skipped because they allegedly don't support our environment.

marun commented:
If we can support the requirements of tests that are currently provider-specific, I'd suggest making the case upstream to change the current provider checks to be something capability-based.  It mirrors my suggestion to make isolation a capability check rather than limiting test execution to a specific network plugin.

danwinship commented:
Updates:
- Added a new commit to flip the default value of `OPENSHIFT_QUIET_OUTPUT` to `true`, because we don't need to be debugging the script itself any more.
- Changed test descriptions/skip messages to sound less contradictory
- Added `skipIfSingleTenant()`/`skipIfMultiTenant()`
- Made `checkPodIsolation()` return an error rather than testing `pluginIsolatesNamespaces()` and checking the result itself
- Changed `multipleNodes bool` to `numNodes int` in `checkPodIsolation()` (leftover from an earlier change to add an `expectSuccess bool` argument rather than calling `pluginIsolatesNamespaces()`, which then got confusing because there were two bools, so I changed the single/multi-node arg to be an int, except then I ended up removing `expectSuccess` and making `checkPodIsolation()` return its error, but I still left the `numNodes` change)
- Changed framework names in services.go
- Changed `launchWebserverService()` return arg name to `serviceAddr`

marun commented:
re-[testonlyextended]

I experienced the same aws provisioning hiccup this morning.

marun commented on test/extended/networking.sh:
```diff
@@ -7,7 +7,7 @@ set -o errexit
 set -o pipefail
 
-if [[ "${OPENSHIFT_QUIET_OUTPUT:-false}" != "true" ]]; then
+if [[ "${OPENSHIFT_QUIET_OUTPUT:-true}" != "true" ]]; then
```
I'm fine with this being the default for non-ci, but having as much output as possible when a job fails is generally useful.  Maybe default to false only if OPENSHIFT_IMAGE_REGISTRY is not provided, since that's now provided by default in ci?

marun replied:
OPENSHIFT_CONFIG_ROOT should probably be considered as well, since ci mode (deploying clusters for each plugin) is only used when that env var is found to be empty.

marun commented on test/extended/networking/util.go:
```diff
@@ -87,22 +88,6 @@ func waitForPodSuccessInNamespace(c *client.Client, podName string, contName str
-	return false
-}
-
 func providerIs(providers ...string) bool {
```
This appears unreferenced after the removal of getMultipleNodes.

marun commented on test/extended/networking/isolation.go:
```diff
@@ -1,35 +1,59 @@
+	var serverNode, clientNode *api.Node
+	serverNode = &nodes.Items[0]
+	if numNodes == 2 {
+		if len(nodes.Items) == 1 {
```
(No action required) Consider adding a skip helper e.g. skipIfSingleNode(nodes).

marun commented on test/extended/networking/services.go:
```diff
@@ -0,0 +1,67 @@
+	})
+})
+
+func checkService(serverFramework, clientFramework *e2e.Framework, numNodes int) error {
```
(No action required) Consider accepting a boolean (e.g. multiNode) instead of an int to indicate whether the check will be performed against multiple nodes.

marun replied:
Sorry, re-read your comment and this is what you intended.

danwinship commented:
- dropped "Make extended networking tests quiet(er) by default" (feel free to do something with it in your branch)
- dropped providerIs()
- (belatedly) renamed checkService() to checkServiceConnectivity()

marun commented:
@danwinship Thank you for your diligence!

@knobunc @eparis @pravisankar @rajatchopra Please merge.

eparis commented:
[merge] with extreme prejudice!

marun commented:
@danwinship I was talking with @liggitt about a failure of a multinode test where multiple nodes were not available, and realized that skipping when the test environment is detected to not support the test is a bad idea.  My first thought was to use a flag like is already being done for isolation, but he pointed out that errors with the target deployment could potentially mask failure conditions.

What do you think of instead using tags in the test name (e.g. [isolated], [non-isolated], [multinode]) to allow test selection via skip regex?

danwinship commented:
> What do you think of switching from flags to using tags in the test name (e.g. [isolated], [non-isolated], [multinode]) to allow test selection via skip regex?

Seems plausible but I haven't really spent enough time with the tests yet to have a good sense of what the right answer is here.

marun commented:
After talking to @jayunit100 I think tags are the way to go. ginkgo's skip regex is arguably preferable because it makes it very obvious to a user of the test binary which tests are going to be run (can be verified with dryRun).  Where runtime checks are required, they could as easily be performed prior to a test run and used to drive the skip regex.  I can't see a reason, then, to make everyone's life more complicated by allowing both skipping by regex and within a test.
